components/tidb_query/src/codec/mysql/time/mod.rs

```diff
@@ -57,605 +48,1156 @@ const MONTH_NAMES_ABBR: &[&str] = &[
                 }
-            } else {
-                if self.fsp > 0 {
-                    let nanos = self.time.nanosecond() / TEN_POW[9 - self.fsp as usize];
-                    format!(
-                        "{}.{1:02$}",
-                        self.time.format("%Y%m%d%H%M%S"),
-                        nanos,
-                        self.fsp as usize
-                    )
+                // If a 'T' is acquired, we expect we have already collected ymd.
+                else if rest[0] == b'T' {
+                    (components.len() == 3).as_option()?;
+                    rest = &rest[1..];
+                }
+                // If a punctuation is acquired, move forward the pointer.
+                else if rest[0].is_ascii_punctuation() {
+                    rest = &rest[1..];
                 } else {
-                    format!("{}", self.time.format("%Y%m%d%H%M%S"))
+                    return None;
                 }
             }
+
+            buffer = rest;
         }
+
+        ((components.len() != 7 && components.len() != 2)
+            || input.as_bytes()[input.len() - components.last().unwrap().len() - 1] == b'.')
+            .as_option()?;
+
+        Some(components)
     }
 
-    fn parse_datetime_format(s: &str) -> Vec<&str> {
-        let trimmed = s.trim();
-        if trimmed.is_empty() {
-            return vec![];
-        }
-        let spes: Vec<&str> = trimmed.split(|c| c < '0' || c > '9').collect();
-        if spes.iter().any(|s| s.is_empty()) {
-            vec![]
+    fn adjust_year(year: u32) -> u32 {
+        if year <= 69 {
+            2000 + year
+        } else if year >= 70 && year <= 99 {
+            1900 + year
         } else {
-            spes
+            year
         }
     }
 
-    fn split_datetime(s: &str) -> (Vec<&str>, &str) {
-        let trimmed = s.trim();
-        if trimmed.is_empty() {
-            return (vec![], "");
+    /// return an array that stores `[year, month, day, hour, minute, second, 0]`
+    fn parse_whole(input: &[u8]) -> Option<[u32; 7]> {
+        let mut parts = [0u32; 7];
+
+        let year_digits = match input.len() {
+            14 | 8 => 4,
+            9..=12 | 5..=7 => 2,
+            _ => return None,
+        };
+
+        parts[0] = bytes_to_u32(&input[..year_digits])?;
+        // If we encounter a two-digit year, translate it to a four-digit year.
+        if year_digits == 2 {
+            parts[0] = adjust_year(parts[0]);
         }
-        let (parts, fracs) = if let Some(i) = trimmed.rfind('.') {
-            (&trimmed[..i], &trimmed[i + 1..])
+
+        for (i, chunk) in input[year_digits..].chunks(2).enumerate() {
+            parts[i + 1] = bytes_to_u32(chunk)?;
+        }
+
+        Some(parts)
+    }
+
+    fn parse_frac(input: &[u8], fsp: u8, round: bool) -> Option<(bool, u32)> {
+        let fsp = usize::from(fsp);
+        let len = input.len();
+
+        let (input, len) = if fsp >= input.len() {
+            (input, len)
```

Comment:
No bugs, please check again. If `fsp >= input.len()`, then we could just ignore the `round`. Otherwise, we should read `fsp + round as usize`s digits.